jgrapht-core/src/main/java/org/jgrapht/graph/interval/IntervalTreeNodeKey.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * (C) Copyright 2003-2018, by Christoph Grüne and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.graph.interval;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Objects;
+
+/**
+ * Implementation of IntervalTreeNodeKey
+ * This is a container class to store the necessary key data for the (augmented) interval tree in the nodes of the tree.
+ *
+ * @param <T> the type of the interval
+ *
+ * @author Christoph Grüne (christophgruene)
+ * @since May 02, 2018
+ */
+public class IntervalTreeNodeKey<T> implements Comparable<IntervalTreeNodeKey<T>> {
+
+    private T key;
+    private Comparator<T> comparator;
```

Comment:
Why does a reference to the comparator have to be stored along with each key?  Isn't the same comparator used for the entire tree?  If that's true, then this class should not exist at all, and you should just use T as the key directly.